build: fix 'threadsafe' feature detection for older gcc

jay opened the pull request:
- Add 'threadsafe' to the feature list shown during build if POSIX threads are being used.

This is a follow-up to 5adb6000 which added support for building a thread-safe libcurl with older versions of gcc where atomic is not available but pthread is.

Reported-by: Dan Fandrich

Fixes https://github.com/curl/curl/issues/12125
Closes #xxxx

dfandrich commented:
I don't know how to easily get cmake going on this ancient FreeBSD, but I tested the autoconf version but it didn't work. The problem is that `HAVE_PTHREAD_H` isn't set as an environment variable anywhere. Replacing that clause with `x"$ac_cv_header_pthread_h" = x"yes"`  (the superfluous `x`s are in there for compatibility with ancient shells that don't handle empty arguments well) fixes it. Thanks!

jay commented:
> I tested the autoconf version but it didn't work. The problem is that `HAVE_PTHREAD_H` isn't set as an environment variable anywhere.

That is confusing because it is supposed to be defined when pthreads is used isn't it?

https://github.com/curl/curl/blob/d755a5f7c009dd63a61b2c745180d8ba937cbfeb/configure.ac#L3741-L3744

> Replacing that clause with `x"$ac_cv_header_pthread_h" = x"yes"` (the superfluous `x`s are in there for compatibility with ancient shells that don't handle empty arguments well) fixes it. Thanks!

Thanks I have incorporated your feedback. Please try the most recently pushed version. I updated both conditionals to prepend the x but I notice in the file it's done both ways. When is one more appropriate than the other? If I understand correctly `test $foo = "bar"` would be a problem if $foo was some command line switch like --qux because then test would interpret it wrong, but in practice when is that actually going to happen with the variables that we set? Or if $foo was unset then wouldn't we have noticed it by now with all the conditionals elsewhere that don't prepend the x?

dfandrich commented:
> That is confusing because it is supposed to be defined when pthreads is used
> isn't it?

Those definitions live in a different namespace and aren't available later in
the configure script. I'll try your change tomorrow.

dfandrich commented:
I tested the update patch with autotools and it works fine. I simulated a missing stdatomic.h in a cmake build on another system and that worked as well.

dfandrich commented:
Just noticed the `&&` in the test command. I'm pretty sure that's a GNUism and should be changed to `-a` for portability.

jay commented:
Thanks. I have added your change, please take another look.

dfandrich commented:
I mis-parsed the statement. It's fine as-is since the shell spawns a second test for the second test clause, but it would avoid a fork if `&& test` is replaced with `-a`.

dfandrich commented:
Yes, that's exactly what I was thinking.

dfandrich commented:
W.r.t. the x's in the `x"$FOO" = xyz` question, yes, we likely would have seen fallout already from all the cases where we don't do that. It's mostly just really, really old shells that have an issue with testing against an empty string. POSIX requires it so it's only pre-POSIX environments where it could make a difference, but since POSIX predates C89 so there could theoretically be a system where it's needed. But, I wouldn't lose too much sleep over it.
